src/uct/base/uct_iface.h

```diff
@@ -161,6 +161,23 @@ typedef struct uct_base_iface {
 UCS_CLASS_DECLARE(uct_base_iface_t, uct_iface_ops_t*,  uct_pd_h, uct_worker_h,
                   const uct_iface_config_t* UCS_STATS_ARG(ucs_stats_node_t*));
 
+/**
+ * Stub interface used for failed endpoints
+ */
+typedef struct uct_failed_iface {
+    uct_iface_t       super;
+    ucs_queue_head_t  pend_q;
+} uct_failed_iface_t;
+
+#define UCT_EP_FAILED(_type, _ep, _iface) \
+    do { \
+        uct_failed_iface_t *f_iface = uct_alloc_failed_iface(_ep, _iface); \
```

Comment:
i think it would be better to add "ucs_class_t *cls" argument to uct_alloc_failed_iface() and call ucs_class_call_cleanup_chain() from that function - so that the macro could be minimal.